### PR TITLE
Terminate foreground tcp and http servers

### PR DESCRIPTION
**Pull request by nartes**

### CODE REVIEW

File: libr/core/cconfig.c [hunk 1]
```diff
@@ -2565,6 +2565,7 @@ R_API int r_core_config_init(RCore *core) {
 	SETPREF ("http.maxport", "9999", "Last HTTP server port");
 	SETPREF ("http.ui", "m", "Default webui (enyo, m, p, t)");
 	SETPREF ("http.sandbox", "true", "Sandbox the HTTP server");
+	SETPREF ("http.no_pipe", "false", "Workaround for slurp error without sandbox");
```
Comment (radare) [inline libr/core/cconfig.c]: config vars dont have lowerdash

File: libr/core/rtr.c [hunk 2]
```diff
@@ -659,26 +659,26 @@ static int r_core_rtr_http_run(RCore *core, int launch, const char *path) {
...
-								free (refstr);
-								ret = -2;
-								goto the_end;
-							}
-						}
-						if (*cmd == ':') {
+
+						if (!strcmp (cmd, "=h*") ||
+						    !strcmp (cmd, "=h--") &&
+						    !r_config_get_i(core->config, "http.sandbox") &&
```
Comment (radare) [inline libr/core/rtr.c]: use r_sandbox_enable(false) instead of assume  the config vars are not lying

File: libr/core/rtr.c [hunk 3]
```diff
@@ -659,26 +659,26 @@ static int r_core_rtr_http_run(RCore *core, int launch, const char *path) {
...
+						    !r_config_get_i(core->config, "cfg.sandbox")) {
+							out = NULL;
+						} else if (*cmd == ':') {
 							/* commands in /cmd/: starting with : do not show any output */
 							r_core_cmd0 (core, cmd + 1);
 							out = NULL;
 						} else {
-							out = r_core_cmd_str_pipe (core, cmd);
+							if (r_config_get_i(core->config, "http.no_pipe") &&
```
Comment (radare) [inline libr/core/rtr.c]: same here. keep it simple

File: libr/core/cconfig.c (hunk 1 above)
Comment (radare) [reply libr/core/cconfig.c]: this should work always, dont add extra config vars, just fix the bug

**Comment by radare** [comment]
ping, please update the pr

File: libr/core/cconfig.c (hunk 1 above)
Comment (nartes) [reply libr/core/cconfig.c]: What's the meaning of command piping? Under the debug tmp file doesn't contain any command. Looks like r_cmd_flush does nothing. I wasn't writing that code. Need some time to understand.

**Comment by nartes** [comment]
No need to update, the bug hasn't been fixed yet. I'm reading the code. I'd rather prefer to have the workaround, rather than to dig into the strange logic.

**Comment by radare** [comment]
what's the problem and how to reproduce it?

**Comment by nartes** [comment]
https://asciinema.org/a/QGJ8TiGZ1KUUiMZP3wxOZVWaI

File: libr/core/rtr.c [hunk 4]
```diff
@@ -690,6 +683,22 @@ static int r_core_rtr_http_run(RCore *core, int launch, const char *path) {
...
+							if (!r_sandbox_enable (0)) {
+								/* do stuff */
+								r_socket_http_close (rs);
+								free (dir);
+								free (refstr);
+								ret = -2;
+								goto the_end;
+							}
+						} else if (!strcmp (cmd, "=h--") &&
+							   !r_sandbox_enable (false)) {
```
Comment (radare) [inline libr/core/rtr.c]: false or 0?

File: libr/core/rtr.c [hunk 5]
```diff
@@ -659,26 +659,19 @@ static int r_core_rtr_http_run(RCore *core, int launch, const char *path) {
...
-								free (dir);
-								free (refstr);
-								ret = -2;
-								goto the_end;
-							}
-						}
-						if (*cmd == ':') {
+
+						if (!strcmp (cmd, "=h*") ||
+						    !strcmp (cmd, "=h--") &&
```
Comment (radare) [inline libr/core/rtr.c]: add one more nesting of parenthesis here around the =-- and the sandbox check or the two strcmps

**Comment by radare** [comment]
ok, ill merge this after the comments fixed. about the str_pipe, i think it can be safely changed to r_core_cmd0 if we wrap it with r_cons_push() / r_cons_pop(). the pipe hack was done by fix this issue before the console buffer was a stack

**Comment by radare** [comment]
please update asap, so we can make it for 2.1

**Comment by nartes** [comment]
Have changed the code according to comments. I've done nothing about piping. I dont' understand push/pop and cmd0 logic in cons module.

I think the PR compiles, but since master's been broken I'd have to rebase it once more. Ping me when master is working again.

**Comment by radare** [comment]
its green

**Pull request merged**